src/ray/util/internal/stream_redirection_utils.h

```diff
@@ -0,0 +1,48 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "ray/util/pipe_logger.h"
+#include "ray/util/scoped_dup2_wrapper.h"
+
+namespace ray::internal {
+
+struct RedirectionHandleWrapper {
+  RedirectionHandleWrapper(MEMFD_TYPE_NON_UNIQUE stream_fd,
+                           const StreamRedirectionOption &opt);
+
+  RedirectionHandleWrapper(const RedirectionHandleWrapper &) = delete;
+  RedirectionHandleWrapper &operator=(const RedirectionHandleWrapper &) = delete;
+  RedirectionHandleWrapper(RedirectionHandleWrapper &&) = default;
+  RedirectionHandleWrapper &operator=(RedirectionHandleWrapper &&) = default;
+
+  // Flush buffered output and restore stream redirection.
+  ~RedirectionHandleWrapper();
+
+  // Flush on redirected stream synchronously.
+  //
+  // TODO(hjiang): Current implementation is naive, which directly flushes on spdlog
+  // logger and could miss those in the pipe; it's acceptable because we only use it in
+  // the unit test for now.
+  void FlushOnRedirectedStream();
+
+  RedirectionFileHandle redirection_file_handle;
+  // Used for restoration.
+  std::unique_ptr<ScopedDup2Wrapper> scoped_dup2_wrapper;
```

Comment:
```suggestion
  RedirectionFileHandle redirection_file_handle_;
  // Used for restoration.
  std::unique_ptr<ScopedDup2Wrapper> scoped_dup2_wrapper_;
```
we use `<name>_` for class members throughout the code (this also belong in a style guide, I know :))

Reply:
Ok a second look, I think it's my bad, it should be `class` instead of `struct`: there's no benefit and necessity to expose these data members.

Reply:
ah yeah sorry I thought it was a class already!